python/ark/module.py

```diff
@@ -191,3 +202,74 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
 
             rt.run()
             return _recursive_ark_to_torch(self.forward_output)
+
+
+class ModuleFn(Function):
+    """
+    Facilitates the integration of ARK modules with PyTorch's
+    autograd system by defining custom forward and backward passes that
+    utilize the user's defined ARK module.
+    """
+    @staticmethod
+    def forward(ctx, input, ark_module):
+        """
+        Returns a PyTorch tensor that is the result
+        of the forward pass of the ARK module.
+        """
+        ark.init()
+        ctx.ark_module = ark_module
+        ark_module.saved_tensors["input"] = input
+        input_ark = Tensor.from_torch(input)
+        output = ark_module.forward(input_ark)
+        with ark.Runtime.get_runtime() as rt:
+            rt.launch(plan=ark.DefaultPlanner().plan())
+            rt.run()
+            output = output.get_torch_view().clone()
+            for name, tensor in ctx.ark_module.saved_tensors.items():
+                ark_module.saved_tensors[name] = (
+                    tensor
+                    if name == "input"
+                    else tensor.get_torch_view().clone()
+                )
+        return output
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        """
+        Converts the gradient outputs to ARK format, computes the gradients for the input
+        and parameters using the ARK module backwards pass, and updates the gradients of the corresponding
+        PyTorch parameters.
+        """
+        ark.init()
+        for name, tensor in ctx.ark_module.saved_tensors.items():
+            if tensor is None or not isinstance(tensor, torch.Tensor):
+                raise ValueError(f"Saved tensor {name} is invalid.")
+            ctx.ark_module.saved_tensors[name] = Tensor.from_torch(tensor)
```

Comment:
Why for saved tensors we store them as PyTorch form and convert them back to ARK form? Why not store ARK form directly